File: javalib/src/main/java/com/strumenta/kolasu/javalib/Traversing.java
package com.strumenta.kolasu.javalib;

import com.strumenta.kolasu.model.BaseASTNode;
import com.strumenta.kolasu.model.Position;
import com.strumenta.kolasu.traversing.ProcessingByPosition;
import com.strumenta.kolasu.traversing.ProcessingStructurally;
import kotlin.jvm.internal.Reflection;
import kotlin.sequences.Sequence;

import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.function.Consumer;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

public class Traversing {

    private static <T> void consumeSequence(Sequence<T> sequence, Consumer<T> consumer) {
        for (Iterator<T> it = sequence.iterator(); it.hasNext();) {
            consumer.accept(it.next());
        }
    }

    private static <T> Stream<T> asStream(Sequence<T> sequence) {
        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(sequence.iterator(),
                Spliterator.ORDERED), false);
    }

    public static Stream<BaseASTNode> walk(BaseASTNode node) {
        return asStream(ProcessingStructurally.walk(node));
    }

    /**
     * Performs a post-order (or leaves-first) node traversal starting with a given node.
     */
    public static Stream<BaseASTNode> walkLeavesFirst(BaseASTNode node) {
        return asStream(ProcessingStructurally.walkLeavesFirst(node));
    }

    public static Stream<BaseASTNode> walkAncestors(BaseASTNode node) {
        return asStream(ProcessingStructurally.walkAncestors(node));
    }

    public static Stream<BaseASTNode> walkDescendantsBreadthFirst(BaseASTNode node) {
        return asStream(ProcessingStructurally.walkDescendants(node, ProcessingStructurally::walk));
    }

    public static Stream<BaseASTNode> walkDescendantsLeavesFirst(BaseASTNode node) {
        return asStream(ProcessingStructurally.walkDescendants(node, ProcessingStructurally::walkLeavesFirst));
    }

    public static <N> Stream<N> walkDescendantsBreadthFirst(BaseASTNode node, Class<N> clazz) {
        return asStream(ProcessingStructurally.walkDescendants(node, Reflection.createKotlinClass(clazz), ProcessingStructurally::walk));
    }

    public static <N> Stream<N> walkDescendantsLeavesFirst(BaseASTNode node, Class<N> clazz) {
        return asStream(ProcessingStructurally.walkDescendants(node, Reflection.createKotlinClass(clazz), ProcessingStructurally::walkLeavesFirst));
    }

    public static void walk(BaseASTNode node, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingStructurally.walk(node), consumer);
    }

    /**
     * Performs a post-order (or leaves-first) node traversal starting with a given node.
     */
    public static void walkLeavesFirst(BaseASTNode node, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingStructurally.walkLeavesFirst(node), consumer);
    }

    public static void walkAncestors(BaseASTNode node, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingStructurally.walkAncestors(node), consumer);
    }

    public static void walkDescendantsBreadthFirst(BaseASTNode node, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingStructurally.walkDescendants(node, ProcessingStructurally::walk), consumer);
    }

    public static void walkDescendantsLeavesFirst(BaseASTNode node, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingStructurally.walkDescendants(node, ProcessingStructurally::walkLeavesFirst), consumer);
    }

    public static <N> void walkDescendantsBreadthFirst(BaseASTNode node, Class<N> clazz, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingStructurally.walkDescendants(node, Reflection.createKotlinClass(clazz), ProcessingStructurally::walk), consumer);
    }

    public static <N> void walkDescendantsLeavesFirst(BaseASTNode node, Class<N> clazz, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingStructurally.walkDescendants(node, Reflection.createKotlinClass(clazz), ProcessingStructurally::walkLeavesFirst), consumer);
    }

    /**
     * Walks the AST within the given position starting from the given node
     * and returns the result as sequence to consume.
     *
     * @param node     the node from which the walk should start
     * @param position the position within which the walk should remain
     */
    public static <N> void walkWithin(BaseASTNode node, Position position, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingByPosition.walkWithin(node, position), consumer);
    }

    /**
     * Walks the AST within the given position starting from each give node
     * and concatenates all results in a single sequence to consume.
     *
     * @param nodes    the nodes from which the walk should start
     * @param position the position within which the walk should remain
     */
    public static <N> void walkWithin(List<BaseASTNode> nodes, Position position, Consumer<BaseASTNode> consumer) {
        consumeSequence(ProcessingByPosition.walkWithin(nodes, position), consumer);
    }

    public static <T> T findAncestorOfType(BaseASTNode node, Class<T> clazz) {
        return ProcessingStructurally.findAncestorOfType(node, clazz);
    }
}

File: core/src/main/kotlin/com/strumenta/kolasu/model/Traversing.kt
package com.strumenta.kolasu.model

import java.util.*
import kotlin.reflect.KClass

/**
 * Some Kotlinization of Deques used as a stack.
 */
typealias Stack<T> = Deque<T>

fun <T> mutableStackOf(): Stack<T> = ArrayDeque()

fun <T> Stack<T>.pushAll(elements: Collection<T>) {
    elements.reversed().forEach(this::push)
}

fun <T> mutableStackOf(vararg elements: T): Stack<T> {
    val stack = mutableStackOf<T>()
    stack.pushAll(elements.asList())
    return stack
}

/**
 * @return walks the whole AST starting from this node, depth-first.
 */
fun Node.walk(): Sequence<Node> {
    val stack: Stack<Node> = mutableStackOf(this)
    return generateSequence {
        if (stack.isEmpty()) {
            null
        } else {
            val next: Node = stack.pop()
            stack.pushAll(next.children)
            next
        }
    }
}

/**
 * @param position the position within which the walk should remain
 * @return walks the AST within the given [position] starting from this node, depth-first.
 */
fun Node.walkWithin(position: Position): Sequence<Node> {
    return if (position.contains(this)) {
        sequenceOf(this) + this.children.walkWithin(position)
    } else if (this.contains(position)) {
        this.children.walkWithin(position)
    } else emptySequence<Node>()
}

/**
 * @param position the position within which the walk should remain
 * @return walks the AST within the given [position] starting from each node
 * and concatenates all results in a single sequence
 */
fun List<Node>.walkWithin(position: Position): Sequence<Node> {
    return this
        .map { it.walkWithin(position) }
        .reduceOrNull { previous, current -> previous + current } ?: emptySequence()
}

/**
 * Performs a post-order (or leaves-first) node traversal starting with a given node.
 */
fun Node.walkLeavesFirst(): Sequence<Node> {
    val nodesStack: Stack<List<Node>> = mutableStackOf()
    val cursorStack: Stack<Int> = ArrayDeque()
    var done = false

    fun nextFromLevel(): Node {
        val nodes: List<Node> = nodesStack.peek()
        val cursor = cursorStack.pop()
        cursorStack.push(cursor + 1)
        return nodes[cursor]
    }

    fun fillStackToLeaf(node: Node) {
        var currentNode: Node = node
        while (true) {
            val childNodes: List<Node> = currentNode.children
            if (childNodes.isEmpty()) {
                break
            }
            nodesStack.push(childNodes)
            cursorStack.push(0)
            currentNode = childNodes[0]
        }
    }
    fillStackToLeaf(this)
    return generateSequence {
        if (done) {
            null
        } else {
            val nodes: List<Node> = nodesStack.peek()
            val cursor = cursorStack.peek()
            val levelHasNext = cursor < nodes.size
            if (levelHasNext) {
                val node: Node = nodes[cursor]
                fillStackToLeaf(node)
                nextFromLevel()
            } else {
                nodesStack.pop()
                cursorStack.pop()
                val hasNext = !nodesStack.isEmpty()
                if (hasNext) {
                    nextFromLevel()
                } else {
                    done = true
                    this
                }
            }
        }
    }
}

/**
 * @return the sequence of nodes from this.parent all the way up to the root node.
 * For this to work, assignParents() must have been called.
 */
fun Node.walkAncestors(): Sequence<Node> {
    var currentNode: Node? = this
    return generateSequence {
        currentNode = currentNode!!.parent
        currentNode
    }
}

/**
 * @return all direct children of this node.
 */
fun Node.walkChildren(): Sequence<Node> {
    return sequence {
        this@walkChildren.properties.forEach { property ->
            when (val value = property.value) {
                is Node -> yield(value)
                is Collection<*> -> value.forEach { if (it is Node) yield(it) }
            }
        }
    }
}

/**
 * @param walker a function that generates a sequence of nodes. By default this is the depth-first "walk" method.
 * For post-order traversal, take "walkLeavesFirst"
 * @return walks the whole AST starting from the childnodes of this node.
 */
@JvmOverloads
fun Node.walkDescendants(walker: (Node) -> Sequence<Node> = Node::walk): Sequence<Node> {
    return walker.invoke(this).filter { node -> node != this }
}

@JvmOverloads
fun <N : Any> Node.walkDescendants(type: KClass<N>, walker: (Node) -> Sequence<Node> = Node::walk): Sequence<N> {
    return walkDescendants(walker).filterIsInstance(type.java)
}

/**
 * Note that type T is not strictly forced to be a Node. This is intended to support
 * interfaces like `Statement` or `Expression`. However, being an ancestor the returned
 * value is guaranteed to be a Node, as only Node instances can be part of the hierarchy.
 *
 * @return the nearest ancestor of this node that is an instance of klass.
 */
fun <T> Node.findAncestorOfType(klass: Class<T>): T? {
    return walkAncestors().filterIsInstance(klass).firstOrNull()
}
